doc/dev/error-categorization.md

```diff
@@ -0,0 +1,167 @@
+Applications cannot handle such errors themselves.
+
+##### Assertion ("C01310")
+
+`Assertion Errors` are such errors which indicate a flaw in the internal logic such as going into a `default`
+branch which you do never expect to happen.
+This category might be used in the future to automatically
+issue a bug to our bugtracker. As of now if you have to use this error please add a message
+indicating that this bug should be reported.
+
+##### Interface ("C01320")
+
+`Interface Errors` errors indicate a wrong usage of Elektra's API. An example would be to pass a NULL pointer to
+`kdbGet`. Also violations of the backend belong into this category. Compared to semantic validation errors,
+this category has its focus on detecting wrong usages of the API instead of a "retry with a different value" approach.
+
+##### Plugin is Broken ("C01330")
+
+`Broken Plugin Errors` errors indicate that a plugin does not behave in an intended way. Unrecognized commands,
+unkown return codes, plugin creation errors, etc. belong to this category. Also uncaught exceptions belong here because
+Elektra expects all exceptions to be caught.
+
+### Conflicting State ("C02000")
+
+`Conflicting State Errors` are errors where the current state is incompatible with the attempted operation.
+Examples are the need for calling `kdbGet` before `kdbSet`. Another example would be to try to push you changes
+into a git repository where the remote branch has already changed.
+These kind of errors are usually in resolver plugins when the state of the file
+has changed without the system knowing. Such errors demand a special synchronization action before retrying.
+
+### Timeout ("C03000")
+
+`Timeout Errors` are easy to categorize as the reaction simply suggests to wait for a short period and retry again.
+Examples are lost connections to a server.
+
+### Validation ("C04000")
+
+`Validation Errors` are heavily used for Elektra's `configuration specification` feature and
+should tell users that their given input does not match a certain pattern/type/expected semantic etc.
+
+Validation errors can either be syntactic or semantic.
+
+#### Syntactic ("C04100")
+
+`Syntactic Errors` are errors which tell users or applications that the current format is not valid.
+Examples are wrong date formats or missing closing brackets `]` inside of a regular expression. Also path related errors
+like missing slashes come into this category. Also parsing errors fall under syntactic errors such as
+unexpected encounters like missing line endings, illegal characters or wrong encodings. Also transformation
+and conversion errors are to be categorizes here because the format of the given input does not allow such
+actions. Users should try a different format and retry pushing it to Elektra.
+
+#### Semantic ("C04200")
+
+`Semantic Errors` are a bit more tricky to solve programmatically. Examples are references to
```

Comment:
Isn't semantic validation now the same as resource errors?

You need a more careful distinction here.

I would say:

- C01100: an essential resource, needed for Elektra to operate is missing (e.g. cannot write to config file)
- C04200: a resource, which is marked essential by the configuration specification is missing (e.g. log file required by application not there)

The handling, however, is usually the same: provide the resource. But if the admin does not want to provide the resource, the handling is different:

- C01100: remount
- C04200: change spec of application OR retry with different value